source/util/move_to_front.h

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_MOVE_TO_FRONT_H_
+#define LIBSPIRV_UTIL_MOVE_TO_FRONT_H_
+
+#include <cassert>
+#include <cstdint>
+#include <ostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace spvutils {
+
+// Log(n) move-to-front implementation. Implements two main functions:
+// IdFromRank - access id by its 1-indexed rank in the move-to-front sequence.
+// RankFromId - get the rank of the given id in the move-to-front sequence.
+// Accessing an id with any of the two functions moves the id to the front
+// of the sequence (rank of 1).
+//
+// The implementation is based on an AVL-based order statistic tree.
+//
```

Comment:
Feedback unaddressed from before:  Should describe the ordering by timestamp; larger timestamps to the left.